### PR TITLE
Allow users to reset passwords

### PR DESCRIPTION
**Pull request by arku**
# 💅🏼 What issue does this fix?

https://github.com/pieforproviders/pieforproviders/issues/121

## 🍂 Before You Submit

* [x] Did you write tests?
* [x] Did you run `bundle exec rspec` from the root?
* [x] Did you run `bundle exec rails rswag` from the root?
* [x] Did you run `bundle exec rubocop` from the root?
* [x] Did you run `yarn lint` in `/client`?
* [x] Did you run `yarn test` in `/client`?
* [ ] Are your primary keys UUIDs on any new tables?

## 🛷 Deployment Considerations

* [ ] Data Migrations
* [ ] Schema Migrations
* [ ] Dependencies

### Confirmed users
1. Sign up for a new account and confirm your email address
2. Go to `/login`
3. Click on the **Reset Password** button
4. Fill in the email address used in Step 1 and click **Reset**
5. You should find the message _You will receive an email with instructions on how to reset your password in a few minutes._
6. Fill in a non-existent email address and click **Reset**
7. You should find the message _Email address not found._
8. Find the password reset instructions email sent to the email address from Step 1 and click on the **Change my password** link.
9. Fill in the form and submit. You should be logged in and redirected to the Getting Started page.
10. If you try to go to `/password/update` or `/password/update?reset_password_token=`, you should be redirected to the login page and the message _Your password reset token is blank. You must click on the reset password link sent to your email address._ should show up
11. If you go to `/password/update?reset_password_token=invalid-token` and submit the form, you should be redirected to the login page and the message _Your password reset token is invalid. Please request a new one._ should show up

### Unconfirmed users

1. Sign up for a new account but don't confirm your email address
2. Go to `/login`
3. Click on the **Reset Password** button
4. Fill in the email address used in Step 1 and click **Reset**
5. You should find the message _You will receive an email with instructions on how to reset your password in a few minutes._
6. Find the password reset instructions email sent to the email address used in Step 1 and click on the **Change my password** link.
7. Fill in the form and submit. You should be redirected to the login page and the message _You have to confirm your email address before continuing._ should appear.

### CODE REVIEW

File: app/mailers/devise_custom_mailer.rb [hunk 1]
```diff
@@ -13,7 +13,7 @@ def confirmation_instructions(record, token, opts = {})
     opts[:subject] = I18n.t('mailers.confirmation_instructions.subject')
     @reply_subject = 'Pie for Providers: question after signup'
```
Comment (weedySeaDragon) [inline app/mailers/devise_custom_mailer.rb:14]: does this need I18n.t ?

**Review by weedySeaDragon — DISMISSED** [review]
The logic is fine, but there's a lot of clean-up/DRYing that should be done with the DeviseMailer and the devise mailer views.
I started to write comments for it all, but decided to stop.  (Mostly because I'm tired.)  I think that cleanup can go in another PR that I could do another time.

Here's a little start

1. devise_custom_mailer should use the mailer layout.  Put the pieforproviders image logo in the layout, then it doesn't have to be repeated in every devise mail view:
   - `app/mailers/devise_custom_mailer.rb`
```
class DeviseCustomMailer < Devise::Mailer
  ...
  layout 'mailer'
```

`views/layouts/mailer` can be changed so that the pieforproviders image is attached:

```


  <body>
    <%= image_tag attachments['pielogo.png'].url, width: '150', align: "middle" %>

    <%= yield %>
  </body>
```

File: app/mailers/devise_custom_mailer.rb (hunk 1 above)
Comment (katelovescode) [reply app/mailers/devise_custom_mailer.rb:14]: No, it's the subject **to us** if they click that email to reply.

**Comment by katelovescode** [comment]
One note, re the testing instructions:
> You should find the message "You will receive an email with instructions on how to reset your password in a few minutes."

Should we give them the "you have to confirm" message HERE, instead of having them go through the process of getting the email, clicking the link, and then being told "j/k you still have to confirm"

File: client/src/i18n/locales/es/index.json [hunk 2]
```diff
@@ -12,7 +12,10 @@
   "passwordRequired": "Se requiere la contraseña",
   "forgotPassword": "¿Olvidaste tu contraseña?",
   "resetPasswordText": "No te preocupes. Haz clic en el botón abajo y revisa tu correo para resetearla.",
+  "resetPasswordInstructions": "Por favor, escriba el correo electrónico utilizado para crear esta cuenta. Haz clic en resetear y comprueba tu bandeja de entrada para obtener más instrucciones.",
```
Comment (csprayregen) [inline client/src/i18n/locales/es/index.json:15]:
```suggestion
  "resetPasswordInstructions": "Por favor, escribe el correo utilizado para crear esta cuenta. Haz clic en resetea y revisa tu buzón para obtener más instrucciones.",
```

File: client/src/i18n/locales/es/index.json [hunk 3]
```diff
@@ -12,7 +12,10 @@
...
   "resetPassword": "Resetea tu contraseña",
+  "reset": "Resetear",
```
Comment (csprayregen) [inline client/src/i18n/locales/es/index.json:17]:
```suggestion
  "reset": "Resetea",
```

File: client/src/i18n/locales/es/index.json [hunk 4]
```diff
@@ -12,7 +12,10 @@
...
+  "cancel": "Cancelar",
```
Comment (csprayregen) [inline client/src/i18n/locales/es/index.json:18]:
```suggestion
  "cancel": "Cancela",
```

File: client/src/i18n/locales/es/index.json [hunk 5]
```diff
@@ -44,6 +47,8 @@
   "passwordInvalid": "La contraseña debe tener un mínimo de 8 caracteres e incluir números y letras",
   "passwordPlaceholder": "Más de 8 caracteres, letras y números",
   "passwordConfirmation": "Confirma la contraseña",
+  "newPassword": "Nueva contraseña",
```
Comment (csprayregen) [inline client/src/i18n/locales/es/index.json:50]:
```suggestion
  "newPassword": "Contraseña nueva",
```

File: client/src/i18n/locales/es/index.json [hunk 6]
```diff
@@ -44,6 +47,8 @@
...
+  "confirmNewPassword": "Confirmar nueva contraseña",
```
Comment (csprayregen) [inline client/src/i18n/locales/es/index.json:51]:
```suggestion
  "confirmNewPassword": "Confirma tu contraseña nueva",
```

File: client/src/i18n/locales/es/index.json [hunk 7]
```diff
@@ -77,8 +84,12 @@
   "genericErrorMessage": "Algo salió mal, inténtalo de nuevo más tarde.",
 
   "alreadyConfirmed": "Este correo electrónico ya ha sido confirmado, por favor intenta iniciar sesión. Si has olvidado tu contraseña, por favor resetéala abajo",
+  "emailUnconfirmed": "Tienes que confirmar tu correo electrónico antes de continuar.",
```
Comment (csprayregen) [inline client/src/i18n/locales/es/index.json:87]:
```suggestion
  "emailUnconfirmed": "Tienes que confirmar tu correo antes de continuar.",
```

File: client/src/i18n/locales/es/index.json [hunk 8]
```diff
@@ -77,8 +84,12 @@
...
   "confirmationPeriodExpired": "Tu periodo de confirmación ha expirado",
   "genericEmailConfirmationError": "Hemos tenido un problema en confirmar tu correo",
+  "emailNotFound": "No se ha encontrado el correo electrónico.",
```
Comment (csprayregen) [inline client/src/i18n/locales/es/index.json:90]:
```suggestion
  "emailNotFound": "No se ha encontrado el correo.",
```

File: client/src/i18n/locales/es/index.json [hunk 9]
```diff
@@ -77,8 +84,12 @@
...
+  "resetPasswordRequestSuccess": "Recibirás un correo electrónico con instrucciones para resetear tu contraseña en unos minutos.",
```
Comment (csprayregen) [inline client/src/i18n/locales/es/index.json:91]:
```suggestion
  "resetPasswordRequestSuccess": "Recibirás un correo con instrucciones para resetear tu contraseña en unos minutos.",
```

File: client/src/i18n/locales/es/index.json [hunk 10]
```diff
@@ -89,6 +100,11 @@
   "contactUs": "Contáctanos",
   "forSupport": "para el soporte técnico",
 
+  "passwordResetTokenBlank": "Tu ficha de resetear la contraseña está en blanco. Debes hacer clic en el enlace de resetear la contraseña enviado a tu correo electrónico.",
```
Comment (csprayregen) [inline client/src/i18n/locales/es/index.json:103]:
```suggestion
  "passwordResetTokenBlank": "Tu token de resetear la contraseña está en blanco. Debes hacer clic en el enlace de resetear la contraseña enviado a tu correo.",
```

File: client/src/i18n/locales/es/index.json [hunk 11]
```diff
@@ -89,6 +100,11 @@
...
+  "passwordResetTokenInvalid": "Tu ficha de resetear la contraseña es inválida. Por favor, solicita una nueva.",
```
Comment (csprayregen) [inline client/src/i18n/locales/es/index.json:104]:
```suggestion
  "passwordResetTokenInvalid": "Tu token de resetear la contraseña es inválido. Por favor, solicita una nueva.",
```

File: client/src/i18n/locales/es/index.json [hunk 12]
```diff
@@ -89,6 +100,11 @@
...
+  "passwordResetTokenExpired": "Tu ficha de resetear la contraseña ha caducado. Por favor, solicita una nueva.",
```
Comment (csprayregen) [inline client/src/i18n/locales/es/index.json:105]:
```suggestion
  "passwordResetTokenExpired": "Tu token de resetear la contraseña ha expirado. Por favor, solicita un token nuevo.",
```

File: config/locales/es.yml [hunk 13]
```diff
@@ -17,8 +19,17 @@ es:
...
       body: ¡Bienvenido a Pie for Providers! Prepárate para aumentar tu porción del pastel con simples herramientas de administración de subsidios de cuidado infantil. Para empezar, por favor verifique su dirección de correo electrónico haciendo clic en el siguiente enlace (también puede copiar y pegar en su navegador).
       confirm_account: Confirmar la cuenta
       questions: "¿Preguntas? Correo electrónico "
-      subject: Verificación del correo electrónico de Pie for Providers
\ No newline at end of file
+      subject: Verificación del correo electrónico de Pie for Providers
+
+    reset_password_instructions:
+      subject: "Pie for Providers - instrucciones para restablecer la contraseña"
+      body: "Alguien ha solicitado un enlace para cambiar su contraseña. Puede hacerlo a través del siguiente enlace."
```
Comment (csprayregen) [inline config/locales/es.yml:29]:
```suggestion
      body: "Alguien ha solicitado un enlace para cambiar tu contraseña. Puedes hacerlo usando el siguiente enlace."
```

File: config/locales/es.yml [hunk 14]
```diff
@@ -17,8 +19,17 @@ es:
...
+      update_password: "Cambiar mi contraseña"
```
Comment (csprayregen) [inline config/locales/es.yml:30]:
```suggestion
      update_password: "Cambia mi contraseña"
```

File: config/locales/es.yml [hunk 15]
```diff
@@ -17,8 +19,17 @@ es:
...
+      note: "Si no has solicitado esto, por favor, ignora este correo electrónico. Tu contraseña no cambiará hasta que accedas al enlace de arriba y crees una nueva."
```
Comment (csprayregen) [inline config/locales/es.yml:31]:
```suggestion
      note: "Si no has solicitado esto, por favor, ignora este correo. Tu contraseña no cambiará hasta que accedas al enlace de arriba y creas una nueva."
```

File: config/locales/es.yml [hunk 16]
```diff
@@ -17,8 +19,17 @@ es:
...
+
+    password_change:
+      subject: "Pie for Providers - contraseña cambiada"
+      body: "Nos contactamos contigo para notificarte que tu contraseña ha sido cambiada."
```
Comment (csprayregen) [inline config/locales/es.yml:35]:
```suggestion
      body: "Contactamos a ti para notificarte que tu contraseña ha sido cambiada."
```

File: app/mailers/devise_custom_mailer.rb [hunk 17]
```diff
@@ -13,7 +13,7 @@ def confirmation_instructions(record, token, opts = {})
...
     @body = I18n.t('mailers.confirmation_instructions.body')
-    @hello = I18n.t('mailers.confirmation_instructions.hello')
+    @hello = I18n.t('hello')
```
[the remark this answers is not in the document]
Comment (arku) [reply app/mailers/devise_custom_mailer.rb:16]: Right. I didn't want to update `DeviseCustomMailer#confirmation_instructions` as it is outside the scope of this PR. Can we do this in a different PR?

**Comment by katelovescode** [comment]
> If you try to go to /password/update?reset_password_token=invalid-token, you should be redirected to the login page and the message Your password reset token is invalid. Please request a new one. should show up

This did not do what I expected.  I visited `http://localhost:3000/password/update?reset_password_token=invalid-token` and got the reset password window.  I tried to log out explicitly to get rid of my JWT token in case having that was somehow overriding the invalid-token, but I still got to the Reset Password page with no alert.

> Fill in the form and submit. You should be redirected to the login page and the message You have to confirm your email address before continuing. should appear.

This worked but one thought for clarity cc @rebeccakarasiktw - should this read "You have successfully changed your password.  You must confirm your email address before logging in."?  Only because it doesn't seem clear if my password was reset or not.

File: config/locales/es.yml [hunk 18]
```diff
@@ -18,7 +20,15 @@ es:
       confirmation_resent: "No se pudo reenviar la confirmación"
   mailers:
     confirmation_instructions:
-      hello: Hola
-      body: ¡Bienvenidas a Pie for Providers! Prepárate para aumentar tu porción del pastel con herramientas sencillas de administración del subsidio. Para empezar, por favor verifica tu dirección de correo haciendo clic en el enlace abajo (también puedes copiar y pegar en su navegador).
```
Comment (katelovescode) [inline config/locales/es.yml:22]: I think this translation shouldn't be changed back to !Bienvenido - the !Bienvenidas is directly from Chelsea, I think.  See resolved comments (last 3 I think) on #208

File: config/locales/es.yml [hunk 19]
```diff
@@ -18,7 +20,15 @@ es:
...
+      body: ¡Bienvenido a Pie for Providers! Prepárate para aumentar tu porción del pastel con simples herramientas de administración de subsidios de cuidado infantil. Para empezar, por favor verifique su dirección de correo electrónico haciendo clic en el siguiente enlace (también puede copiar y pegar en su navegador).
       questions: "¿Preguntas? Correo electrónico "
-      subject: Verificación del correo de Pie for Providers
\ No newline at end of file
+      subject: Verificación del correo electrónico de Pie for Providers
```
Comment (katelovescode) [inline config/locales/es.yml:25]: Ditto here, see #208 comments

File: spec/cypress/app_commands/scenarios/reset_password_token_used.rb [hunk 20]
```diff
@@ -0,0 +1,4 @@
+# frozen_string_literal: true
+
+user = User.first
+user&.update! reset_password_token: nil
```
Comment (katelovescode) [inline spec/cypress/app_commands/scenarios/reset_password_token_used.rb:4]: 😍

File: spec/cypress/support/on-rails.js [hunk 21]
```diff
@@ -1,25 +1,30 @@
...
+      return response.body
+    })
 })
 
 Cypress.Commands.add('app', function (name, command_options) {
-  return cy.appCommands({name: name, options: command_options}).then((body) => {
-    return body[0]
-  })
+  return cy
+    .appCommands({ name: camelToSnakeCase(name), options: command_options })
```
Comment (katelovescode) [inline spec/cypress/support/on-rails.js:20]: Could you put in a comment about why this is necessary?

**Review by katelovescode — CHANGES_REQUESTED** [review]
Really really minor comments, but this looks great and it works.  Nice job.

File: spec/cypress/support/on-rails.js (hunk 21 above)
Comment (arku) [reply spec/cypress/support/on-rails.js:23]: Done.

File: config/locales/es.yml (hunk 18 above)
Comment (arku) [reply config/locales/es.yml:22]: Fixed.

File: config/locales/es.yml (hunk 19 above)
Comment (arku) [reply config/locales/es.yml]: Fixed.

File: client/src/PasswordReset/PasswordReset.js [hunk 22]
```diff
@@ -0,0 +1,156 @@
+import React, { useEffect, useState } from 'react'
+import { useHistory, useLocation } from 'react-router-dom'
+import { useTranslation } from 'react-i18next'
+import { Form, Input } from 'antd'
+import { PaddedButton } from '_shared/PaddedButton'
+import { useApiResponse } from '_shared/_hooks/useApiResponse'
+
+export const PasswordReset = () => {
```
[the remark this answers is not in the document]
Comment (arku) [reply client/src/PasswordReset/PasswordReset.js]: Done.

File: client/src/ErrorAlert/ErrorAlert.js [hunk 23]
```diff
@@ -0,0 +1,51 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import ResendToken from 'ResendToken'
+import { useTranslation } from 'react-i18next'
+
+const contactUs = ({ t, message }) => {
+  return `${message} <a href="mailto:tech@pieforproviders.com">${t(
```
[the remark this answers is not in the document]
Comment (arku) [reply client/src/ErrorAlert/ErrorAlert.js]: Done.

**Review by weedySeaDragon — APPROVED** [review]
code looks good.

**Review by katelovescode — APPROVED** [review]
Nice work! Please add testing instructions to the ticket itself and move it to QA

**Pull request merged**